design/20241124-rename-bunde-to-clusterbundle.md

```diff
@@ -0,0 +1,162 @@
+the controller is able to change owner references on target resources from `Bundle` to `ClusterBundle` without
+facing issues.
+
+We must also ensure that the `ClusterBundle` reconciler isn't deleting target resources it thinks are "orphaned".
+We believe the controller will only delete target resources if/when the controller resource exists, and
+otherwise rely on the Kubernetes garbage collector to delete orphans. But this should be tested/verified.
+
+### Temporary Technical Controller
+
+When the functional controller is migrated to target `ClusterBundle`, we want to introduce a temporary technical
+controller targeting the now deprecated `Bundle` resource.
+
+This controller should just create or update the `ClusterBundle` with the same name as the reconciled `Bundle`.
+Since the APIs should have identical `spec`, this should be a simple controller.
+
+To allow a user to migrate their resources from `Bundle` to `ClusterBundle`, the controller should not act
+on `Bundle` delete events. Nor add an owner reference to `ClusterBundle` to avoid the Kubernetes garbage collector
+to act on deletes.
+
+For reconciling a `ClusterBundle` from `Bundle` we should investigate the possibility of "smart field manager name logic".
+It should be possible to read the field manager name used (by user) to create/update the `Bundle` resource.
+If we use the same field manager name in controller (per resource) to create/update `ClusterBundle`, we think it's less
+likely that users will have to force conflicts when migrating from `Bundle` to `ClusterBundle`.
+
+### Risks and Mitigations
+
+#### Target configmaps/secrets are accidentally deleted
+
+Since using owner references in this project, we need to be extra careful when performing changes like this.
+
+We mitigate this risk by **NOT** cascading deletes of `Bundle` to `ClusterBundle`.
+This is done by **NOT** adding owner references to `ClusterBundle`, and instead use the resource `name`
+to map between `ClusterBundle` and `Bundle` resources.
+
+Owner references on target resources are changed by the reconciler from `Bundle` to `ClusterBundle`,
+which means targets could only be deleted if the `ClusterBundle` is deleted or `namespaceSelector`
+is changed. This is the expected behavior.
+
+#### Target configmaps/secrets appearing as orphaned
+
+This could happen if a users deletes a `Bundle` resource expecting the target configmaps/secrets
+to vanish. Unfortunately this won't happen by the defensive approach taken to not cascade `Bundle`
+delete to `ClusterBundle`.
+
+In order to achieve the wanted behavior, a user would have to delete the corresponding `ClusterBundle`,
+which appears like a simple workaround that we should document.
+
+## Design Details
+
+### Test Plan
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+Downgrading from a version where the migration has started will probably not be possible and should
+be clearly documented and stated in the release notes.
+
+## Alternatives
```

Comment:
question: Is it worth listing the alternative of "don't try to migrate anyone"?

Like, if we marked `Bundle` as deprecated but continued to support a controller for it, it wouldn't cause any problems beyond a bit more code to maintain, right?

(I'm not necc. saying that we _should_ do that, but it seems like a scenario we should address in this doc right?)

(I'm assuming that we change the group to `trust-manager.io` for `ClusterBundle`, which would mean that keeping `trust.cert-manager.io Bundle` wouldn't conflict with a future namespace scoped `trust-manager.io Bundle`)

Reply:
I have tried to explain above why I **don't** think we should support both resources independently - at the same time. But I will add this option to the alternatives for sure.

Renaming the API group is excellent input! I will add this too.

Reply:
New `trust-manager.io` API group is now added to the list of goals. 👍 

And added a new independent controller as one of the alternatives.